src/main/java/com/synopsys/integration/detect/ApplicationUpdater.java

```diff
@@ -624,12 +642,62 @@ private File handleResponse(Response response, String currentInstalledVersion, F
         } else if (response.getStatusCode() == HttpStatus.SC_NOT_MODIFIED) {
             logger.info("{} Present Detect installation is up to date - skipping download.", LOG_PREFIX);
         } else {
+            String problemUrl = captureProblemDetectUrl(downloadUrl);
             String message = StringUtils.isNotBlank(response.getStatusMessage()) ? response.getStatusMessage() : EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getStatusCode(), Locale.ENGLISH);
-            logger.warn("{} Unable to download artifact. Response code: {} {}", LOG_PREFIX, response.getStatusCode(), message);
+            logger.warn("{} Unable to download artifact from {}. Response code: {} {}", LOG_PREFIX, problemUrl, response.getStatusCode(), message);
         }
         return null;
     }
     
+    /**
+     * This method attempts to determine the true URL that is hosting Detect as we
+     * are often redirected from the BlackDuck hosted URL.
+     * 
+     * @param downloadUrl the BlackDuck hosted URL to download detect from, ie:
+     *                    https://localhost/api/tools/detect
+     * @return the true URL that is hosting the Detect jar after following the 302
+     *         redirect from /api/tools/detect
+     * @throws IOException
+     */
+    private String captureProblemDetectUrl(HttpUrl downloadUrl) throws IOException {
+        String problemUrl = downloadUrl.toString(); 
+        
+        // We need to build a new client to communicate with BlackDuck. This is because the main client we use to
+        // talk to BlackDuck will follow 302 redirects and we will be unable to determine and report on where the download
+        // actually failed from.
+        try {
+            HostnameVerifier hostnameVerifier;
+            SSLContext sslContext;
+
+            if (trustCertificate) {
+                sslContext = SSLContextBuilder.create().loadTrustMaterial(new TrustAllStrategy()).build();
+                hostnameVerifier = new NoopHostnameVerifier();
+            } else {
+                sslContext = SSLContexts.createDefault();
+                hostnameVerifier = SSLConnectionSocketFactory.getDefaultHostnameVerifier();
+            }
+            SSLConnectionSocketFactory connectionFactory = new SSLConnectionSocketFactory(sslContext, hostnameVerifier);
+
+            CloseableHttpClient instance = HttpClients.custom().setSSLSocketFactory(connectionFactory)
+                    .disableRedirectHandling().build();
+
+            HttpGet httpGet = new HttpGet(downloadUrl.uri());
+            CloseableHttpResponse response = instance.execute(httpGet);
+
+            Header locationHeader = response.getFirstHeader("location");
+            
+            if (locationHeader != null) {
+                problemUrl = locationHeader.getValue();
```

Comment:
Not sure how likely this would be to happen, but what if the "true" location is several redirects away? Should we consider doing this recursively?

Reply:
Also, a more general thought: it's not necessarily clear to me that users would prefer to know the final URL rather than the one that the client started with.

If I imagine myself debugging issues based on the log, I might want to know the initial URL. That way I could reproduce the exact HTTP redirect path using cURL or something.

Reply:
And yet another thought: if the response is not a redirect but a real response with the Detect jar, the download could take an additional while which is not a great user experience. The download should ideally fail quickly if it does.

Reply:
I believe the URL this code is calling is pretty specialized and will be called only when we have issues downloading. So to get Detect we will always call /api/tools/detect initially. I believe what that redirects to is configurable in the BlackDuck UI so I think the first location header we get will be what we are after and want to log. So in the case where I say to BlackDuck I want to self-host at URL https://xyz.com we'll see output like:

2023-10-10 14:20:20 EDT INFO  [main] --- Detect-Self-Updater:  Checking https://localhost/api/tools/detect API for centrally managed Detect version to download to /Users/dterry/tmp.
2023-10-10 14:20:22 EDT WARN  [main] --- Detect-Self-Updater:  Unable to download artifact from https://xyz.com. Response code: 200 OK

So in this case the client knows what they started with and what it initially redirected to (what they put in the BlackDuck UI). I was thinking that was enough for a simple debug function. If it's a real response with a Detect jar then we should never invoke this code.

Reply:
Thanks for the additional information! I am more on board with the change the way it is now.